src/mame/drivers/segas18_astormbl.cpp

```diff
@@ -0,0 +1,759 @@
+	PORT_DIPSETTING(    0x00, DEF_STR( On ) )
+	PORT_DIPNAME( 0x1c, 0x1c, DEF_STR( Difficulty ) )   PORT_DIPLOCATION("SW2:3,4,5")
+	PORT_DIPSETTING(    0x04, DEF_STR( Easiest ) )
+	PORT_DIPSETTING(    0x08, DEF_STR( Easier ) )
+	PORT_DIPSETTING(    0x0c, DEF_STR( Easy ) )
+	PORT_DIPSETTING(    0x1c, DEF_STR( Normal ) )
+	PORT_DIPSETTING(    0x10, DEF_STR( Hard ) )
+	PORT_DIPSETTING(    0x14, DEF_STR( Harder ) )
+	PORT_DIPSETTING(    0x18, DEF_STR( Hardest ) )
+	PORT_DIPSETTING(    0x00, "Special" )
+	PORT_DIPNAME( 0x20, 0x20, "Coin Chutes" )       PORT_DIPLOCATION("SW2:6")
+	PORT_DIPSETTING(    0x20, "3" )
+	PORT_DIPSETTING(    0x00, "1" )
+	PORT_DIPUNUSED_DIPLOC( 0x40, IP_ACTIVE_LOW, "SW2:7" )
+	PORT_DIPUNUSED_DIPLOC( 0x80, IP_ACTIVE_LOW, "SW2:8" )
+INPUT_PORTS_END
+
+
+
+TILE_GET_INFO_MEMBER(segas18_astormbl_state::get_text_tile_info)
+{
+	int tile_number = m_textram[tile_index];
+
+	tileinfo.set(0,
+			(tile_number & 0x1ff),
+			(tile_number >> 9) % 8,
+			0);
+
+	tileinfo.category = (tile_number >> 15) & 1;
+}
+
+void segas18_astormbl_state::sys16_textram_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	COMBINE_DATA(&m_textram[offset]);
+	m_text_layer->mark_tile_dirty(offset);
+}
+
+void segas18_astormbl_state::video_start()
+{
+	m_text_layer = &machine().tilemap().create(
+		*m_gfxdecode,
+		tilemap_get_info_delegate(*this, FUNC(segas18_astormbl_state::get_text_tile_info)),
+		TILEMAP_SCAN_ROWS,
+		8, 8,
+		64, 28);
+
+	m_text_layer->set_transparent_pen(0);
+	m_text_layer->set_scrollx(512 - 320);
+}
+
+
+void segas18_astormbl_state::draw_tile(screen_device& screen, bitmap_ind16& bitmap, const rectangle &cliprect, int tilenum, int colour, int xpos, int ypos, uint8_t pri, int transpen, bool opaque)
+{
+	gfx_element* gfx = m_gfxdecode->gfx(0);
+	bitmap_ind8 &priority_bitmap = screen.priority();
+
+	const uint8_t* gfxdat = gfx->get_data(tilenum);
+
+	for (int y = 0; y < 8; y++)
+	{
+		int realypos = ypos + y;
+
+		for (int x = 0; x < 8; x++)
+		{
+			int realxpos = xpos + x;
+
+			if ((realypos >= cliprect.min_y) && (realypos <= cliprect.max_y) && (realxpos >= cliprect.min_x) && (realxpos <= cliprect.max_x))
+			{
+				uint16_t* dst = &bitmap.pix(realypos);
+				uint8_t* pridst = &priority_bitmap.pix(realypos);
+				uint8_t pix = gfxdat[(y*8) + x];
+
+				if (pix || opaque)
+				{
+					dst[realxpos] = pix | (colour * 8);
+					pridst[realxpos] = pri;
+				}
+			}
+		}
+	}
+}
+
+void segas18_astormbl_state::draw_layer(screen_device& screen, bitmap_ind16& bitmap, const rectangle &cliprect, int layer, bool opaque, int pri0, int pri1)
+{
+	for (int quadrant = 0; quadrant < 4; quadrant++)
+	{
+		int xbase = 0, ybase = 0;
+
+		if (quadrant & 2)
+			xbase = 0;
+		else
+			xbase = 512;
+
+		xbase += 0x140;
+
+		if (quadrant & 1)
+			ybase = 256;
+		else
+			ybase = 0;
+
+		if (layer == 0)
+			ybase ^= 256;
+
+		int y = 0;
+
+		int yscroll = layer & 1 ? m_tileyscroll0[0] : m_tileyscroll1[0];
+
+		for (int i = 0; i < 0x20; i++)
```

Comment:
I might just be tired, but is there a reason for having separate `y` and `i` variables here?  They seem to do the same thing.  At any rate, even if `y` is necessary, it could be declared/incremented in the `for` loop’s control expressions.

Reply:
just leftover from the 'work out how this works' discovery process, will clean up